src/compiler/types.ts

```diff
@@ -449,6 +450,14 @@ namespace ts {
         ThisNodeOrAnySubNodesHasError = 1 << 17, // If this node or any of its children had an error
         HasAggregatedChildData = 1 << 18, // If we've computed data from children and cached it in this node
 
+        // This flag will be set to true when the parse encounter dynamic import so that post-parsing process of module resolution
+        // will not walk the tree if the flag is not set. However, this flag is just a approximation because once it is set, the flag never get reset.
+        // (hence it is named "possiblyContainDynamicImport").
+        // During editing, if dynamic import is remove, incremental parsing will *NOT* update this flag. This will then causes walking of the tree during module resolution.
+        // However, the removal operation should not occur often and in the case of the removal, it is likely that users will add back the import anyway.
+        // The advantage of this approach is its simplicity. For the case of batch compilation, we garuntee that users won't have to pay the price of walking the tree if dynamic import isn't used.
+        possiblyContainDynamicImport = 1 << 19,

```

Comment:
I think we need to find another way to do this to avoid adding the expense for scenarios like:
```ts
import
(function () {})();
```
Here I was in the middle of adding a new import statement and now must pay the cost of walking the whole tree on every incremental parse. 

We should instead consider tracking a sorted array of `ImportCall` expressions on `SourceFile` during parse. For an incremental parse, we can reuse `ImportCall` nodes from the array that are both reusable and outside of the change range, and insert any new `ImportCall` nodes we parse. 

As a result, even if we do add an `ImportCall` to the document, we don't incur the cost of a full walk in either of the initial or incremental parse scenarios. We can also consider doing this for `require` calls to avoid the full walk in those cases as well.